### DISCUSSION TITLE
How to compile in Windows 11

### DISCUSSION THREAD

**Discussion by carrots66** [burningalchemist/sql_exporter#470] [Q&A] [answered]
Hello, I am running out of ideas. Have installed GNU Make and Go 1.22. 

No matter how I set the env variables I get this kind of message:
_go: GOPATH entry is relative; must be absolute path: "C".
For more details see: 'go help gopath'
make: *** [drivers-minimal] Error 2_

That was trying to build the drivers minimal, if doing a "make build" with go 1.21 installed and GNU Make I get:
_C:\go-work\sql_exporter-master>make build
'uname' is not recognized as an internal or external command,
operable program or batch file.
process_begin: CreateProcess(NULL, uname -m, ...) failed.
'GOOS' is not recognized as an internal or external command,
operable program or batch file.
make: *** [promu] Error 1_

Can anybody shed some light on how to troubleshoot any of these?

**Comment by burningalchemist** [comment]
Hi @carrots66, you need to have configured Golang on your system. I guess you can follow the answer from Stackoverflow here - https://stackoverflow.com/questions/35884965/gopath-error-in-windows-gopath-entry-is-relative-must-be-absolute-path

**Reply by carrots66** [reply]
Thanks a lot for your prompt reply and your work on the exporter.
I trust my path is fine, an absolute one. Tried many combinations already to no avail.

![image](https://github.com/burningalchemist/sql_exporter/assets/164333207/ef529628-e0f0-4937-8514-e4e490126c83)

**Reply by burningalchemist** [reply]
could you drop the output of `go env` command?

**Reply by carrots66** [reply]
C:\go-work\sql_exporter-master>go env
set GO111MODULE=
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\ASDASD\AppData\Local\go-build
set GOENV=C:\Users\ASDASD\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=C:\go-work\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\go-work
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=C:\golang
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLCHAIN=auto
set GOTOOLDIR=C:\golang\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.21.8
set GCCGO=gccgo
set GOAMD64=v1
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=0
set GOMOD=C:\go-work\sql_exporter-master\go.mod
set GOWORK=
set CGO_CFLAGS=-O2 -g
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-O2 -g
set CGO_FFLAGS=-O2 -g
set CGO_LDFLAGS=-O2 -g
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=C:\Users\ASDASD\AppData\Local\Temp\go-build4284481960=/tmp/go-build -gno-record-gcc-switches

**Reply by carrots66** [reply]
These are my other paths set, which I trust are fine. Installed vainilla GNU Make, have no CYGWIN (info just in case).
![image](https://github.com/burningalchemist/sql_exporter/assets/164333207/78d6d4ab-839d-41b6-b9aa-1239e56b7272)

**Reply by burningalchemist** [reply]
Yeah, I guess this might be related since in the Makefile we do have some GOPATH system override which corresponds to linux-like envs:

`GOPATH := $(firstword $(subst :, ,$(shell $(GO) env GOPATH)))`

Maybe you can adjust (e.g. remove these lines) lines 15-16 in the Makefile? I believe I didn't have any issues with Windows 10, but I can't really find any Windows 11 box around. Also used bash-driven configurations not native ones.

Give it a try and let me know what's the output.

**Comment by carrots66** [comment]
Makes no difference. For the shake of clarity: the path error only happens when trying to "make drivers-minimal", all other options complain about uname and GOOS. On the other hand I don't see promu on the go bin directory.
I will try on a W10 machine and see what gives. Thanks again.
![image](https://github.com/burningalchemist/sql_exporter/assets/164333207/afb25235-a43d-43f5-9e50-ac5290d12fd1)

**Reply by burningalchemist** [reply]
I think it depends on the environment you execute these commands on as `uname` is also from the coreutils package (linux).

Let me try to install it with native Make and we can turn it into a recipe. :)

**Reply by burningalchemist** [reply]
To be fair this file hasn't been changed for years, so it might need some refresh as `sql_exporter` seems to become popular haha. :)

**Reply by carrots66** [reply]
The flexibility that your version gives us is outworldly, of course it should be popular :).
We want to compile so as to add the capability to read encrypted .yamls.

**Reply by burningalchemist** [reply]
I usually integrate `sql_exporter` with tools like [SOPS](https://github.com/getsops/sops) for encrypted secrets, but yeah I fully understand the intention, especially if there's no option to wire several tools together and so on. :+1:

I'll have a look today as promised, and most likely update the Makefile - seems like we need to detect OS and run specific commands to set up the env.

The gist is that we need to have `promu` utility installed. In fact you can try the following mitigation:

- install promu from [here](https://github.com/prometheus/promu) with this command - `go install github.com/prometheus/promu@v0.14.0`
- run `promu build` from the working directory (where `.promu.yml` exists)

- for `drivers.go` generation you might want to manually run the commands from `drivers` target in the Makefile. You just need to substitute the argument - `minimal`, `all` or `custom`.

Check it out, this might help, just to unblock you.

**Reply by carrots66** [reply]
promu tips did not help situation. I tried on a my W10 box at home and more or less got the same results.
I managed to compile it on a linux box after updating the go version and added decrypting capabilities to the config and collector read functions. I only need it to compile into exe now.
Will look into crosscompiling it which I guess should be possible but if you sort that Makefile that'd be greatest.
Thanks for all the replies

**Reply by burningalchemist** [reply]
Hey @carrots66, gotcha! Could you share what was the error following the promu steps?
Yeah, I'll do it today after working hours, stay tuned. 👍🎶

**Reply by burningalchemist** [reply]
Hey @carrots66, I've created branch `win-makefile`, but I need to carefully test it to avoid any regressions before the merge. In the meantime, if you could test it on your end, that would be great. Check it out and try to run `make` commands again.

**Comment by carrots66** [comment]
Hello,

Sorry for late reply. I managed to cross-compile in linux and that worked
just fine.

However, I have tried the promu steps on the Widnows machine, for some
reason it was installing the promu.exe over the c:\go-work\go\bin directory
rather than the go installation in program files.
I copied the .exe over the installation at c:\program files\go\bin and
promu build compiled in Windows just fine too.

I will test the win makefile it in Windows 10 and 11 and report back.

**Comment by carrots66** [comment]
This is what I get on Windows 10 (home computer)
![image](https://github.com/burningalchemist/sql_exporter/assets/164333207/54160109-3d29-4160-a16e-36841f8c9a7f)

On Windows 11 (work computer) I have connectivity issues, so most of my
problems seem to come from not being able to download the requirements.
Will see if I can progress further on that.

**Reply by burningalchemist** [reply]
I guess the image wasn't attached to the message, so could you please reattach it? :)

**Reply by carrots66** [reply]
done, sorry about that

will also type down output just in case

_C:\go-work\sql_exporter-fix-win-makefile\sql_exporter-fix-win-makefile>make build
process_begin: CreateProcess(NULL, pwd, ...) failed.
">> building binaries"
promu: error: expected argument for flag '--prefix', try --help
make: *** [build] Error 1_

**Reply by burningalchemist** [reply]
No worries. Alright, we're closer, the `promu` is executed, the prefix is not passed, though. I'll fix and ask you to try again. :)

**Reply by burningalchemist** [reply, marked as answer]
@carrots66 

Ok, I've managed to fix and test it on my windows laptop, it should work. Could you please try and test it again? For `build` and `drivers-%` it should work just fine for all the cases now.

I'll create a pull request to merge it to master.

`style` and `vet` targets still use unix utilities like `grep` and `find`, so they won't work on Windows out of the box. It's possible to install these as GNU utilities, but it'd be a user responsibility. The commands are still more developer-oriented, so I don't think that's relevant anyway.

**Comment by carrots66** [comment]
Hello @burningalchemist 

Worked perfectly fine. Many thanks for that! :)
![image](https://github.com/burningalchemist/sql_exporter/assets/164333207/57ba6c71-0ab9-4730-81ef-3e8c8a08a1d9)

**Reply by burningalchemist** [reply]
Perfect! It's merged to master, by the way. 👍 And thanks for reporting the issue. 🙂
